### PR TITLE
DEPR: `Styler.where` possible deprecation

### LINKED ISSUE #40124: DEPR: `Styler.where` in favour of using `applymap` and `apply`
`Styler` by definition is a conditional renderer. The functions it provides, `apply` and `applymap`, serve as conditional functions to render a style.

The function `Styler.where` therefore, in the first case, seems unnecessary and indirectly confusing since it is a conditional rendering function implying perhaps the other methods are not.  Not only is it just a 'convenience' wrapper of `applymap`:

```
def where(cond, value, other, subset, **kwargs):
    return self.applymap(lambda v: value if cond(v) else other, subset=subset, **kwargs)
```
But I also think, in the second case, it is quite inconvenient for a couple reasons:
1)  The docs are ambiguous:
```
cond: callable: cond should take a scalar and return a boolean.
**kwargs: dict: pass along to cond.
```
and in fact `kwargs` appear in the code but do nothing.

2) It feels that there is more benefit to be had by a user learning how you use `apply` and `applymap` to the same effect through good examples.

For example: 
`styler.applymap(lambda v: props if cond(v) else None, subset=subset)` or
`styler.apply(lambda d: np.where(d > 10, props, ""), axis=None, subset=subset)`

seem pretty straightforward replacements???

Perhaps documenting in the first instance that `apply` and `applymap` are preferred over `where` is non-controversial.

### PR DESCRIPTION
**Pull request by attack68**
closes #40124 

Proposal to deprecate the method: `Styler.where`, since `Styler.applymap` and `Styler.where` are so similar in function, and return signatures, that it is almost confusing to have 2 methods for the same thing.

Note that `Styler.where` offers no advantage over `applymap` since internally it just restructures the callable and uses `applymap` itself:

```
def where(..):
    return self.applymap(lambda val: value if cond(val, **kwargs) else other, subset=subset)
```

### CODE REVIEW

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -1080,10 +1078,12 @@ def where(
         Updates the HTML representation with a style which is
         selected in accordance with the return value of a function.
 
+        .. deprecated:: 1.3.0
+
         Parameters
         ----------
         cond : callable
-            ``cond`` should take a scalar and return a boolean.
+            ``cond`` should take a scalar, and optional args, and return a boolean.
```
Comment (attack68) [inline pandas/io/formats/style.py:1086]: this was conflicting with line 1095 in the docstring:
```
**kwargs : dict
            Pass along to ``cond``.
```

**Comment by jreback** [comment]
i would actually deprecate .applymap, as its a non-vectorized version of .where (not that this matters for Styler generally).

ok can you separate out the bug fix and do the deprecation in a different PR? (where can discuss)

**Comment by attack68** [comment]
> i would actually deprecate .applymap, as its a non-vectorized version of .where (not that this matters for Styler generally).
> 
> ok can you separate out the bug fix and do the deprecation in a different PR? (where can discuss)

my other PR #40845 fixes the bug. This one I left as the potential deprecation. `Styler.where` is not vectorized, it's effectively a one-liner invoking `applymap`.

**Comment by jreback** [comment]
cc @pandas-dev/pandas-core any thoughts here

**Comment by rhshadrach** [comment]
To use `where` in order to accomplish `applymap` seems unnatural:

    df.where(cond=lambda _: True, some_style)

To use `applymap` to accomplish `where` seems natural:

    df.applymap(lambda x: some_style if cond(x))

**Comment by dsaxton** [comment]
> To use `where` in order to accomplish `applymap` seems unnatural:
> 
> ```
> df.where(cond=lambda _: True, some_style)
> ```
> 
> To use `applymap` to accomplish `where` seems natural:
> 
> ```
> df.applymap(lambda x: some_style if cond(x) else other_style)
> ```

I think I agree with this. As a user I always find myself having to stop and think what `where` actually does, on the other hand methods like `map`, `apply`, and `applymap` are more obvious.

**Comment by jreback** [comment]
ok thanks, yeah no harm in deprecating

**Pull request merged**